File: front/apps/admin-react/src/api/business/sysResourceMenu.ts
import API from '../index'

export type ResourceMenuType = 'CATALOG' | 'MENU' | 'BUTTON' | 'EMBEDDED' | 'LINK'

export interface ResourceMenuMetadata {
  icon?: string
  order?: number
  hidden?: boolean
  authorities?: string[]
  [key: string]: unknown
}

export interface ResourceMenu {
  id: number
  parentID?: number
  treePath?: string
  menuType: ResourceMenuType
  path: string
  redirect: string
  alias: string
  name: string
  component: string
  metadata: ResourceMenuMetadata
  sortOrder: number
  isEnabled: boolean
  remark: string
  children?: ResourceMenu[]
  canWrite?: boolean
  canDelete?: boolean
}

export interface ResourceMenuNode {
  id: number
  parentID?: number
  menuType: ResourceMenuType
  path: string
  redirect: string
  name: string
  component: string
  icon: string
  order: number
  sortOrder?: number
  hidden: boolean
  authorities: string[]
  isUrl: boolean
  children?: ResourceMenuNode[]
}

export interface ReqResourceMenuCreate {
  parentID?: number
  menuType: ResourceMenuType
  path: string
  redirect: string
  alias: string
  name: string
  component: string
  metadata: ResourceMenuMetadata
  sortOrder: number
  isEnabled: boolean
  remark: string
}

export interface ReqResourceMenuUpdate extends Partial<ReqResourceMenuCreate> {
  id: number
}

export interface ReqResourceMenuBatchDelete {
  ids: number[]
}

function menuList(req: PagingRequest) {
  return API.Post<Res<PagingResult<ResourceMenu>>>('/api/sys/resource/menu/list', req, {
    cacheFor: 0,
  })
}

function menuTree() {
  return API.Get<Res<ResourceMenuNode[]>>('/api/sys/resource/menu/tree', {
    cacheFor: 0,
  })
}

async function menuCreate(req: ReqResourceMenuCreate) {
  await API.Post<Res>('/api/sys/resource/menu/create', req, {
    cacheFor: 0,
  }).send()
}

async function menuUpdate(req: ReqResourceMenuUpdate) {
  await API.Post<Res>('/api/sys/resource/menu/update', req, {
    cacheFor: 0,
  }).send()
}

async function menuDel(req: ReqResourceMenuBatchDelete) {
  await API.Post<Res>('/api/sys/resource/menu/del', req, {
    cacheFor: 0,
  }).send()
}

export const ResourceMenuApi = {
  menuList,
  menuTree,
  menuCreate,
  menuUpdate,
  menuDel,
}

File: backend/admin/internal/services/orm/models/sys_resource_menu.go
package models

import (
	"gorm.io/plugin/soft_delete"
	"orm-crud/gormc/mixin"
)

func init() {
	Models = append(Models, &SysResourceMenu{})
}

// SysResourceMenu 对应表 sys_resource_menu
type SysResourceMenu struct {
	mixin.AutoIncrementID
	mixin.CreatedAt
	mixin.UpdatedAt
	mixin.OperatorID
	mixin.Remark
	mixin.SortOrder
	mixin.Metadata
	mixin.IsEnabled
	DeletedAt soft_delete.DeletedAt `gorm:"column:deleted_at;softDelete:milli;not null;default:0;index" json:"deletedAt"`
	MenuType  string                `gorm:"column:menu_type;type:varchar(32);not null;default:'MENU';comment:菜单类型 CATALOG: 目录 MENU: 菜单 BUTTON: 按钮 EMBEDDED: 内嵌 LINK: 外链" json:"menuType"`
	Path      string                `gorm:"column:path;type:varchar(1024);not null;comment:路径，类型为按钮时为操作名" json:"path"`
	Redirect  string                `gorm:"column:redirect;type:varchar(1024);not null;default:'';comment:重定向地址" json:"redirect"`
	Alias     string                `gorm:"column:alias;type:varchar(255);not null;default:'';comment:路由别名" json:"alias"`
	Name      string                `gorm:"column:name;type:varchar(255);not null;default:'';comment:路由命名" json:"name"`
	Component string                `gorm:"column:component;type:varchar(255);not null;default:'';comment:前端页面组件" json:"component"`

	// 简单树结构字段（保留父级关系与路径）
	ParentID *uint64 `gorm:"column:parent_id;type:bigint;comment:父级ID" json:"parentID"`
	TreePath *string `gorm:"column:tree_path;type:varchar(1024);comment:节点路径" json:"treePath"`

	ParentSysResourceMenu *SysResourceMenu `gorm:"foreignKey:ParentID;references:ID" json:"parentSysResourceMenu"`
}

func (SysResourceMenu) TableName() string {
	return "sys_resource_menu"
}

type CatalogMeta struct {
	Icon   string `json:"icon"`
	Order  int32  `json:"order"`
	Hidden bool   `json:"hidden"`
}

type MenuMeta struct {
	Icon   string `json:"icon"`
	Order  int32  `json:"order"`
	Hidden bool   `json:"hidden"`
}

type ButtonMeta struct {
	Authorities []string `json:"authorities"`
}

type EMBEDDEDMeta struct {
	Authorities []string `json:"authorities"`
}

type LinkMeta struct {
	Authorities []string `json:"authorities"`
}
